src/ResolveIteratorCapableTrait.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Traversable;
+use Iterator;
+use IteratorAggregate;
+use Exception as RootException;
+use OutOfRangeException;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+
+/**
+ * Capable of resolving an iterable to its innermost iterable.
+ *
+ * @since [*next-version*]
+ */
+trait ResolveIteratorCapableTrait
+{
+    /**
+     * Finds the deepest iterator that matches.
+     *
+     * Because the given traversable can be an {@see IteratorAggregate}, it will try to get its inner iterator.
+     * On each iterator, it will run the test function.
+     * It will keep going inwards until one of these conditions occurs:
+     *      - The iterator matches (test function returns `true`).
+     *          This can also be the supplied iterator. In this case the iterator will be returned.
+     *      - The limit or the maximal depth are reached.
+     *          In this case, if the iterator is not a match, an exception will be thrown.
+     *
+     * @since [*next-version*]
+     *
+     * @param Traversable $iterator The iterator to resolve.
+     * @param callable    $test     The test function which determines when the iterator is considered to be resolved.
+     *                              Default: Returns `true` on first found instance of {@see Iterator}.
+     * @param $limit int|float|string|Stringable The depth limit for resolution.
+     *
+     * @throws InvalidArgumentException If limit is not a valid integer representation.
+     * @throws OutOfRangeException      If infinite recursion is detected, or the iterator could not be resolved within the depth limit.
+     *
+     * @return Iterator The inner-most iterator, or whatever the test function allows.
+     */
+    protected function _resolveIterator(Traversable $iterator, $test = null, $limit = null)
+    {
+        $origIterator = $iterator;
+
+        // Default test function
+        if (!is_callable($test)) {
```

Comment:
I would also make this more strict. If `null`, use default. If not `null` and not a `callable`, throw.

Reply:
Otherwise, in the event of a bug in consuming code, this would "fail" silently if a non-callable argument is given.